examples/client/tui/commands.nim

```diff
@@ -788,6 +792,48 @@ proc command*(self: Tui, command: CallRpc) {.async, gcsafe,
   except:
     self.wprintFormatError(getTime().toUnix, "invalid arguments.")
 
+# SendTransaction --------------------------------------------------------------
+
+proc help*(T: type SendTransaction): HelpText =
+  let command = "sendtransaction"
+  HelpText(command: command, aliases: aliased[command], parameters: @[
+    CommandParameter(name: "from", description: "The address for the sending account"),
+    CommandParameter(name: "to", description: "The destination address of the message, left empty for a contract-creation transaction."),
+    CommandParameter(name: "value", description: "The value transferred for the transaction in wei"),
+    CommandParameter(name: "maxPriorityFee", description: "Max priority fee in wei"),
+    CommandParameter(name: "maxFee", description: "Max fee in wei"),
+    CommandParameter(name: "gasLimit", description: "The amount of gas to use for the transaction (unused gas is refunded)."),
+    CommandParameter(name: "data", description: "ABI byte string containing the data of the function call on a contract, or contract-creation initialisation code."),
+    CommandParameter(name: "nonce", description: "Integer of the nonce"),
+    CommandParameter(name: "password", description: "Account password")
+    ], description: "Sends an EIP1559 transaction")
+
+proc new*(T: type SendTransaction, args: varargs[string]): T =
+  let fromAddress = eth_common.parseAddress(args[0])
+  let transaction = Transaction(
+    txType: TxType.TxEip1559,
+    to: some(eth_common.parseAddress(args[1])),
+    value: args[2].u256,
+    maxPriorityFee: args[3].parseBiggestInt,
+    maxFee: args[4].parseBiggestInt,
+    gasLimit: args[5].parseBiggestInt,
+    payload: args[6].hexToSeqByte,
+    nonce: uint64(args[7].parseBiggestInt)
```

Comment:
@emizzle, @michaelsbradleyjr  what's the best way to validate that the arguments have the correct format? Here I receive a combination of eth address, integers, and hex strings

Reply:
Good question, made me realize something that I missed when reviewing your PR for the `/call` command:

https://github.com/status-im/nim-status/blob/feat/send-transaction/examples/client/tui/common.nim#L44-L45
> all fields on types that derive from Command should be of type `string`

The way the command parser is currently implemented, neither a command's `split` nor `new` procs are in a good position to do validation and if they raise the program will crash. See this comment at the top of the commands module:

https://github.com/status-im/nim-status/blob/feat/send-transaction/examples/client/tui/commands.nim#L18-L30

Commonly, that means letting a field for a missing parameter be an empty string, and with all fields being strings and minimal logic in those parts, the command parser is unlikely to crash.

The best place to do validation and convert from string to non-string is therefore in the `command` proc, and looking at what you have so far, it seems you're going that direction.

I suggest creating a:
```nim
proc validate(command: SendTransaction): SendTransactionResult
```

And then in the command proc you can do something like:
```nim
let cmdResult = SendTransaction.validate()

if cmdResult.isErr:
  self.wprintFormatError(getTime().toUnix, "Invalid arguments: " & cmdResult.error)
else:
  let cmd = cmdResult.get
  asyncSpawn self.client.sendTransaction(
    eth_common.parseAddress(cmd.fromAddress),
    Transaction(
      txType: TxType.TxEip1559,
      to: some(eth_common.parseAddress(cmd.to)),
      value: cmd.value.u256,
      maxPriorityFee: cmd.maxPriorityFee.parseBiggestInt,
      maxFee: cmd.maxFee.parseBiggestInt,
      gasLimit: cmd.gasLimit.parseBiggestInt,
      payload: cmd.payload.hexToSeqByte,
      nonce: uint64(cmd.nonce.parseBiggestInt),
    ),
    cmd.password
)
```

I will understand if you find it a bit awkward. I think the way the command parser works can be improved in a future PR/s. For example,  in addition to `help`, `new`, `split`, and `command` procs for each command type, there could be a `validate` command and it would automatically get used by the [parser](https://github.com/status-im/nim-status/blob/feat/send-transaction/examples/client/tui/parser.nim). If `validate` results in an `isErr` then the parser would take care of printing the error, instead of that logic having to be implemented by hand in each `command` proc.

That still leave the awkwardness re: converting from string to non-string, and often times there is overlap between validation logic and conversion logic. I need to think more about the relationship between the parser and dispatch and how we can make that part as pain-free as possible.

Reply:
Hmm, I guess for now, the `validate` proc for `SendTransaction` doesn't have to return a `SendTransaction` instance wrapped in `Result`. The instance wrapped in `Result` can instead be of the type that has the fields converted as you want.

So the last part would be:
```nim
asyncSpawn self.client.sendTransaction(cmd.fromAddress, cmd.transaction, cmd.password)
```

Reply:
That approach will work generally, but wouldn't lend itself to being plugged into the parser because of how the types have to match up for dispatch.

I'll think about this some more, there's probably a fun and clever way to make it pretty clean with the help of one/more macros and maybe some `cast`ing.

Reply:
Alright, I did the change although I did the validation in place instead of having a separate proc (to not deviate much from what is done in the other commands). In this PR I also changed the CallRpc command so the params are a string instead of a Json